pandas/tseries/index.py

```diff
@@ -681,6 +687,67 @@ def _add_delta(self, delta):
             result = result.tz_convert(self.tz)
         return result
 
+    def _add_offset(self, offset):
+        from pandas import Timedelta
```

Comment:
I would move this entire thing to the actual offsets themselves. E.g add a method in ``DateOffset``, maybe ``period_delta(DatetimeIndex i)``, e.g. it takes a ``DatetimeIndex`` (and for the ones that don't support, either raise a ``ValueError``. Add the ``_begend_offset`` (would split into 2 methods), add as an internal method of the offset as well.